app/src/main/java/net/gsantner/markor/format/AutoFormatter.java

```diff
@@ -0,0 +1,368 @@
+            result = indent + String.format("%s%c ", getNextOrderedValue(oLine.value), oLine.delimiter);
+        } else if (uLine.isUnorderedOrCheckList && uLine.lineEnd != uLine.groupEnd && _dend >= uLine.groupEnd) {
+            String itemPrefix = uLine.newItemPrefix;
+            result = indent + itemPrefix;
+        } else {
+            result = indent;
+        }
+
+        return result;
+    }
+
+    public static class PrefixPatterns {
+        public final Pattern prefixUnorderedList;
+        public final Pattern prefixCheckBoxList;
+        public final Pattern prefixOrderedList;
+
+        public PrefixPatterns(Pattern prefixUnorderedList, Pattern prefixCheckBoxList, Pattern prefixOrderedList) {
+            this.prefixUnorderedList = prefixUnorderedList;
+            this.prefixCheckBoxList = prefixCheckBoxList;
+            this.prefixOrderedList = prefixOrderedList;
+        }
+    }
+
+    public static class ListLine {
+        protected static final int INDENT_DELTA = 2;
+
+        protected final CharSequence text;
+
+        public final int lineStart, lineEnd;
+        public final String line;
+        public final int indent;
+        public final boolean isEmpty;
+        public final boolean isTopLevel;
+
+        public ListLine(CharSequence text, int position) {
+
+            this.text = text;
+            lineStart = StringUtils.getLineStart(text, position);
+            lineEnd = StringUtils.getLineEnd(text, position);
+            line = text.subSequence(lineStart, lineEnd).toString();
+            indent = StringUtils.getNextNonWhitespace(text, lineStart) - lineStart;
+            isEmpty = (lineEnd - lineStart) == indent;
+            isTopLevel = indent <= INDENT_DELTA;
+        }
+
+        // Empty lines are children of any line level
+        public boolean isParentLevelOf(final ListLine line) {
+            return line.isEmpty || (!isEmpty && (line.indent - indent) > INDENT_DELTA);
+        }
+
+        public boolean isChildLevelOf(final ListLine line) {
+            return isEmpty || (!line.isEmpty && (indent - line.indent) > INDENT_DELTA);
+        }
+
+        public boolean isSiblingLevelOf(final ListLine line) {
+            return !isParentLevelOf(line) && !isChildLevelOf(line);
+        }
+    }
+
+    /**
+     * Class to parse a line of text and extract useful information
+     */
+    public static class OrderedListLine extends ListLine {
+        // TODO parametrize?
+        private static final int FULL_GROUP = 2;
+        private static final int VALUE_GROUP = 3;
+        private static final int DELIM_GROUP = 4;
+
+        private final PrefixPatterns prefixPatterns;
+
+        public final boolean isOrderedList;
+        public final char delimiter;
+        public final int numStart, numEnd;
+        public final int groupStart, groupEnd;
+        public final String value;
+
+        public OrderedListLine(CharSequence text, int position, PrefixPatterns prefixPatterns) {
```

Comment:
Why are we passing the search patterns around? Is it because zim and markdown have different Ordered list etc formats and you want to select behavior by passing in the pattern-set.

If so, I would suggest use 'format' or 'type' in the name. `class FormatPatterns` or something like that to show that this is format specific.

Reply:
Yes exactly, it's a small helper wrapper class I introduced which is used exclusively to distinguish the prefixes between Markdown and Zim, to avoid long argument lists.
I like to keep this name for now because the wrapped patterns only refer to prefixes, but I'd change this if other format-specific patterns should be included in the future.